File: demos/tinystories/research/generate.cpp
#include <torch/torch.h>
#include <torch/script.h>
#include <iostream>
#include <fstream>
#include <string>
#include <unordered_map>
#include <vector>
#include <algorithm>
#include <random>
#include "json.hpp"

using json = nlohmann::json;

const int MAX_LENGTH = 500;
const int SOS_IDX = 1;
const int EOS_IDX = 2;

std::unordered_map<std::string, int> stoi;
std::vector<std::string> itos;

void load_vocab(const std::string& path) {
    std::ifstream f(path);
    json vocab_json;
    f >> vocab_json;
    stoi = vocab_json["stoi"].get<std::unordered_map<std::string, int>>();
    itos = vocab_json["itos"].get<std::vector<std::string>>();
    // std::cout << "Vocab size: " << stoi.size() << " words\n";
	// std::cout << "First few tokens:\n";
	// for (int i = 0; i < std::min(10, (int)itos.size()); ++i) {
	//     std::cout << i << ": " << itos[i] << '\n';
	// }
}

int argmaxDecode(const torch::Tensor& scores) {
    return scores.argmax().item<int>();
}

int sampleDecode(const torch::Tensor& scores, double temperature = 0.5) {
    torch::Tensor scaled = scores / temperature;
    torch::Tensor probs = torch::softmax(scaled, 0);
    torch::Tensor sampled = torch::multinomial(probs, 1);
    return sampled.item<int>();
}

int nucleusDecode(const torch::Tensor& scores, double p = 0.9, double temperature = 0.5) {
    torch::Tensor scaled = scores / temperature;
    torch::Tensor probs = torch::softmax(scaled, 0);

    auto sorted = probs.sort(-1, /*descending=*/true);
    auto sorted_probs = std::get<0>(sorted);
    auto sorted_indices = std::get<1>(sorted);
    auto cumulative_probs = torch::cumsum(sorted_probs, 0);

    int cutoff = (cumulative_probs < p).sum().item<int>();
    torch::Tensor topk_indices = sorted_indices.slice(0, 0, cutoff + 1);

    torch::Tensor filtered_probs = torch::zeros_like(probs);
    filtered_probs.index_put_({topk_indices}, probs.index({topk_indices}));
    filtered_probs /= filtered_probs.sum();

    return torch::multinomial(filtered_probs, 1).item<int>();
}

torch::Tensor tokenize(const std::string& prompt) {
    torch::Tensor tokens = torch::zeros({1, MAX_LENGTH}, torch::kLong);
    tokens[0][0] = SOS_IDX;
    int idx = 1;
    std::istringstream iss(prompt);
    std::string word;
    while (iss >> word && idx < MAX_LENGTH) {
        tokens[0][idx++] = stoi.count(word) ? stoi[word] : stoi["<UNK>"];
    }
    return tokens;
}

void generate_text(
    torch::jit::script::Module& model,
    const torch::Tensor& src,
    int prompt_len,
    const std::string& prompt,
    std::function<int(const torch::Tensor&)> decode_fn,
) {
    torch::Tensor input = src.clone();
    std::cout << "\n\n" << prompt;
    for (int t = prompt_len + 1; t < MAX_LENGTH; ++t) {
        auto output = model.forward({input}).toTensor();
        auto next_scores = output[0][t - 1]; // [vocab_size]
        int idx = decode_fn(next_scores);
        input[0][t] = idx;
        std::string word = itos[idx];

        if (word == "<EOS>") break;
        if (word.find_first_of(".,!?\"'") == std::string::npos)
            word = " " + word;
        std::cout << word << std::flush;
    }
    std::cout << "\n";
}

int main() {
    torch::manual_seed(0);
    torch::globalContext().setDeterministicCuDNN(true);

    load_vocab("vocab.json");
    torch::jit::script::Module model = torch::jit::load("model.pt");
    model.eval();

    while (true) {
        std::string prompt;
        std::cout << "\n\nPrompt:\n";
        std::getline(std::cin, prompt);
        if (prompt.empty()) continue;

        torch::Tensor src = tokenize(prompt).to(torch::kCPU);
        int prompt_len = prompt.empty() ? 0 : std::count_if(prompt.begin(), prompt.end(), ::isspace) + 1;

        std::cout << "\n\n[Argmax Sampling]";
        generate_text(model, src, prompt_len, prompt, argmaxDecode);

        std::cout << "\n\n[Temperature Sampling]";
        generate_text(model, src, prompt_len, prompt,
            [](const torch::Tensor& s) { return sampleDecode(s, 0.5); });

        std::cout << "\n\n[Nucleus Sampling]";
        generate_text(model, src, prompt_len, prompt,
            [](const torch::Tensor& s) { return nucleusDecode(s, 0.9, 0.5); });
    }

    return 0;
}

File: demos/tinystories/research/generate.py
import torch
torch.use_deterministic_algorithms(True)
import sys
from models.TransformerLM import *
from data.TinyStories import *
from spacy.tokenizer import Tokenizer
torch.serialization.add_safe_globals([Vocabulary, Tokenizer])
from torch.distributions import Categorical
torch.backends.cudnn.deterministic = True

use_cuda_if_avail = True
if use_cuda_if_avail and torch.cuda.is_available():
    device = "cuda"
else:
    device = "cpu"

MAX_LENGTH = 500

# print(device)

def main():
   CHKPT_PATH = "./chkpts/2ZJPbu_TinyStories"
   chkpt = torch.load(CHKPT_PATH, map_location=torch.device('cpu'), weights_only=False)
   config = chkpt['config']

   print(CHKPT_PATH +" // "+str(chkpt['epoch']))
   # load vocab
   vocab = chkpt['vocab']

   # load model
   model = TransformerLM(len(vocab), config["d_model"], config["n_heads"], config["n_layers"])
   model.load_state_dict(chkpt['model_state_dict'])
   model.to(device)

   
   while True:
    # ask for prompt
    prompt = input("\n\nPrompt:\n")

    # numeralize prompt
    num_prompt = vocab.text2idx(prompt)
    l = len(num_prompt)

    
    for sampler in [argmaxDecode, sampleDecode, nucleusDecode]:
      torch.manual_seed(0)
      random.seed(0)
      torch.cuda.manual_seed(0)
      torch.cuda.manual_seed_all(0)
      
      src = torch.zeros(1,MAX_LENGTH)
      src[0,0] = 1 # <SOS>
      src[0,1:l+1] = torch.Tensor(num_prompt)
      src = src.to(dtype=int, device=device)
      print("\n\n")
      print(sampler)
      print(prompt, end="",flush=True)
      for t in range(l+1, MAX_LENGTH):
          out = model(src)


          src[0,t] =  sampler(out[:,t-1,:])
          
          w = vocab.idx2text([src[0,t].cpu().item()])[0]

          if w == "<EOS>":
              break
          if not any(x in w for x in [".",",","\"","'","!","?"]):
              w = " "+w
          
          print(w,  end='',flush=True)
      
      print("\n")
   sys.exit(1)


def argmaxDecode(scores):
    idx = torch.argmax(scores, 1)
    return idx

def sampleDecode(scores, temp = 0.5):
    scores = scores.squeeze()/temp
    dist = Categorical(torch.nn.functional.softmax(scores, dim=0))
    idx = dist.sample()
    return idx

def nucleusDecode(scores, p=0.9, temp = 0.5):
    scores = scores.squeeze()/temp
    probs = torch.nn.functional.softmax(scores, dim=0)

    sorted_probs, sorted_indices = torch.sort(probs, descending=True)
    cumulative_probs = torch.cumsum(sorted_probs, dim=0)

    cutoff = torch.searchsorted(cumulative_probs, p)

    topk_indices = sorted_indices[:cutoff + 1]
    filtered_probs = torch.zeros_like(probs)
    filtered_probs[topk_indices] = probs[topk_indices]
    filtered_probs /= filtered_probs.sum()

    dist = Categorical(filtered_probs)
    return dist.sample()


main()